tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/openwire/SecurityOpenWireTest.java

```diff
@@ -16,70 +16,675 @@
+         } catch (JMSSecurityException e) {
+            fail("Should be able to create sender to given FQQN: " + e.getMessage());
+         }
+
+         try {
+            session1.createProducer(destination2);
+            fail("Should not be able to create sender to given FQQN: ");
+         } catch (JMSSecurityException e) {
+            // Expected: should fail to create sender to this FQQN destination
+         }
+
+         try {
+            session2.createProducer(destination2);
+            // Expected: should be able to create sender to allowed destination
+         } catch (JMSSecurityException e) {
+            fail("Should be able to create sender to given FQQN: " + e.getMessage());
+         }
+
+         try {
+            session2.createProducer(destination1);
+            fail("Should not be able to create sender to given FQQN: ");
+         } catch (JMSSecurityException e) {
+            // Expected: should fail to create sender to this FQQN destination
+         }
+      }
+   }
+
+   @Test
+   public void testAnonymousSendersWithFineGrainedAuthToFQQNAsQueues() throws Exception {
+      doTestAnonymousSenderWithFineGrainedAuthToFQQNs(false);
+   }
+
+   @Test
+   public void testAnonymousSendersWithFineGrainedAuthToFQQNAsTopics() throws Exception {
+      doTestAnonymousSenderWithFineGrainedAuthToFQQNs(true);
+   }
+
+   private void doTestAnonymousSenderWithFineGrainedAuthToFQQNs(boolean topic) throws Exception {
+      try (Connection connection1 = factory.createConnection(ALLOWED_USER, PASS);
+           Connection connection2 = factory.createConnection(ALLOWED_USER_ALTERNATE, PASS)) {
+
+         final Session session1 = connection1.createSession(false, Session.AUTO_ACKNOWLEDGE);
+         final Session session2 = connection2.createSession(false, Session.AUTO_ACKNOWLEDGE);
+
+         final Destination destination1;
+         final Destination destination2;
+
+         if (topic) {
+            destination1 = session1.createTopic(FQQN);
+            destination2 = session2.createTopic(FQQN_ALTERNATE);
+         } else {
+            destination1 = session1.createQueue(FQQN);
+            destination2 = session2.createQueue(FQQN_ALTERNATE);
+         }
+
+         final MessageProducer producer1 = session1.createProducer(null);
+         final MessageProducer producer2 = session2.createProducer(null);
+
+         producer1.setDeliveryMode(DeliveryMode.PERSISTENT); // ALLOWED USER
+         producer2.setDeliveryMode(DeliveryMode.PERSISTENT); // ALTERNATE ALLOWED USER
+
+         try {
+            producer1.send(destination1, session1.createMessage());
+            // Expected: should send to allowed destination
+         } catch (JMSSecurityException e) {
+            fail("Should be able to send to given FQQN: " + e.getMessage());
+         }
+
+         try {
+            producer1.send(destination2, session1.createMessage());
+            fail("Should not be able to send to given FQQN: ");
```

Comment:
The failure message ending ": " will suggest something is missing. Maybe drop that, or else add a destination getName() to provide the value.

Same thing in various other spots.